File: lab6/C++/src/Equipment.cpp
#include <stdexcept>
#include "Equipment.hpp"
#include "WarehouseEntry.hpp"
#include "Enums.hpp"

Equipment::Equipment(int id, MaterialType materialType, const std::string& name, int count)
    : WarehouseEntry<int>(id, name, materialType)
{
    if (count < Equipment::MIN_COUNT)
        throw std::invalid_argument("Недопустимое количество");
    m_count = count;
}

int Equipment::GetCount() const {
    return m_count;
}

bool Equipment::IsEmpty() const {
    return m_count <= 0;
}

void Equipment::Increase(int count) {
    m_count += count;
}

int Equipment::Decrease(int count) {
    int taken = std::min(m_count, count);
    m_count -= taken;
    return taken;
}

File: lab8/Java/src/warehouse/equipments/Equipment.java
package warehouse.equipments;

import enums.MaterialType;
import request.SupplyRequestDetail;
import warehouse.WarehouseEntry;

public abstract class Equipment extends WarehouseEntry<Integer> {

    int count;

    public static final Integer MIN_COUNT = 1;

    public Equipment(int id, String name, MaterialType materialType, int count) { 
        super(id, name, materialType);
        if(count < MIN_COUNT){
            throw new IllegalArgumentException("Недопустимое количество");
        }         
        this.count = count;
    }

    public int getCount(){
        return count;
    }
    @Override
    public void increase(Number value){
        count+=value.intValue();
    }

    @Override
    public Integer decrease(Number value){
        int taken = Math.min(count, value.intValue());
        count-=taken;
        return taken;
    }
    
    @Override
    public Boolean isEmpty(){
        return count < MIN_COUNT;
    }

    @Override
    public abstract Boolean isMatches(SupplyRequestDetail detail);
}